stdlib/src/commonMain/kotlin/it/unibo/collektive/stdlib/spreading/GradientCast.kt

```diff
@@ -187,12 +189,29 @@ inline fun <reified ID : Any, reified Value, reified Distance : Comparable<Dista
                         // Since the source collection is sorted, the subcollections will be sorted as well
                         pathsBySource.getOrPut(it.path.source) { mutableListOf() }.apply { add(it) }
                     }
-                    while (size < maxPaths) {
-                        val candidates = pathsBySource.values
-                            .mapNotNull { it.removeFirstOrNull() }
-                            .take(maxPaths - size - fromLocalSource.size)
-                            .map { it.toLocalPath(accumulateData) }
-                        addAll(candidates)
+                    when {
+                        pathsBySource.size == 1 -> {
+                            // If there is only one source, take as many paths as necessary from there
+                            val toAdd = pathsBySource.values.single()
+                                .asSequence()
+                                .take(toTake)
+                                .map { it.toLocalPath(accumulateData) }
+                            addAll(toAdd)
+                        }
+                        else -> {
+                            // If there are multiple sources, alternate to preserve information
+                            while (size < toTake) {
+                                val toIterate = pathsBySource.values.iterator()
+                                while (toIterate.hasNext() && size < toTake) {
+                                    val pathsForSource = toIterate.next()
+                                    val path = pathsForSource.removeFirst()
+                                    if (pathsForSource.isEmpty()) {
+                                        toIterate.remove()
+                                    }
+                                    add(path.toLocalPath(accumulateData))
+                                }
+                            }
+                        }
```

Comment:
This code is very difficult to understand.
Is there a way to refactor this intricate logic into separate, simpler functions?

Reply:
I need performance here. I can do a minimal refactoring, but I won't do anything that slows the computation down (I already have performance issues)

Reply:
See if you like it now